src/modules/sharing/guards/sharing-permissions.guard.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  ForbiddenException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { User } from '../../user/user.domain';
+import { UserUseCases } from '../../user/user.usecase';
+import { FolderAttributes } from '../../folder/folder.attributes';
+import { verifyWithDefaultSecret } from '../../../lib/jwt';
+import { SharingService } from '../../sharing/sharing.service';
+import {
+  SharedWithType,
+  SharingActionName,
+} from '../../sharing/sharing.domain';
+import {
+  PermissionsMetadataName,
+  PermissionsOptions,
+} from './sharing-permissions.decorator';
+import { Workspace } from '../../workspaces/domains/workspaces.domain';
+import { WorkspaceTeam } from '../../workspaces/domains/workspace-team.domain';
+import { Folder } from '../../folder/folder.domain';
+
+@Injectable()
+export class SharingPermissionsGuard implements CanActivate {
+  constructor(
+    private reflector: Reflector,
+    private userUseCases: UserUseCases,
+    private sharingUseCases: SharingService,
+  ) {}
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    const permissionsOptions: PermissionsOptions = this.reflector.get(
+      PermissionsMetadataName,
+      context.getHandler(),
+    );
+    const { action } = permissionsOptions;
+
+    const request = context.switchToHttp().getRequest();
+    const requester = request?.user as User;
+
+    if (!requester) {
+      return false;
+    }
+
+    const resourcesToken = request.headers['internxt-resources-token'];
+
+    if (!resourcesToken || typeof resourcesToken !== 'string') {
+      return true;
+    }
+
+    const decoded = verifyWithDefaultSecret(resourcesToken) as
+      | {
+          owner?: {
+            uuid?: User['uuid'];
+          };
+          sharedRootFolderId?: FolderAttributes['uuid'];
+          sharedWithType: SharedWithType;
+          workspace?: {
+            workspaceId: Workspace['id'];
+            teamId: WorkspaceTeam['id'];
+          };
+        }
+      | string;
+
+    if (typeof decoded === 'string') {
+      throw new ForbiddenException('Invalid token');
+    }
+
+    const userIsAllowedToPerfomAction =
+      await this.isWorkspaceMemberAbleToPerformAction(
+        requester,
+        decoded.sharedRootFolderId,
+        action,
+      );
```

Comment:
I only added support for sharing between members, but we can adapt this guard so it works with individual sharings and teams in a future. My thought is use the decoded token to know if the sharing the user is navigating through is for individuals, teams or members and make the respective checks according to that.